js/ui/grid_core/ui.grid_core.editing.js

```diff
@@ -828,15 +840,25 @@ const EditingController = modules.ViewController.inherit((function() {
 
             const changes = this.getChanges();
             changes.forEach(change => {
-                const key = change.key;
+                const isInsert = change.type === DATA_EDIT_DATA_INSERT_TYPE;
 
-                if(isDefined(key) && change.type === DATA_EDIT_DATA_INSERT_TYPE) {
-                    const loadedRowIndex = this._getLoadedRowIndexByInsertKey(items, e, key);
+                if(!isInsert) {
+                    return;
+                }
 
-                    const item = this._generateNewItem(key);
-                    if(loadedRowIndex >= 0 && this._needInsertItem(change, changeType, items, item)) {
-                        items.splice(key.dataRowIndex ? loadedRowIndex : 0, 0, item);
-                    }
+                let key = change.key;
+                let insertKey = this._getInternalData(key)?.insertKey;
+                if(isInsert && (!isDefined(change.key) || !isDefined(insertKey))) {
+                    const keys = this._getInsertKey();
+                    change.key = keys.key;
+                    key = keys.key;
+                    insertKey = keys.insertKey;
+                }
```

Comment:
isInsert is redundant in the following condition:
```suggestion
                if(!isDefined(change.key) || !isDefined(insertKey)) {
                    const keys = this._getInsertKey();
                    change.key = keys.key;
                    key = keys.key;
                    insertKey = keys.insertKey;
                }
```
Would you please clarify what scenario is covered by this code? 

In addition, insertKey and key that are stored in an internal data item are confusing. Maybe rename insertKey to insertId or insertInfo or something else?

Reply:
User can add his own changes to the init configuration. So we need to process it and add key/insertKey.